### PR TITLE
Feature data destination

### PR DESCRIPTION
**Pull request by TobiKattmann**
Adds the possibilty to prescribe the destination filename for the addData function. This is useful e.g. for a deformed mesh is written in an evaluation and given a different name than the base mesh. In the foloowing evaluation this deformed output mesh is symlinked with the base-mesh name such that the cfg does not need to be changed.

PR to pcarruscag/FADO that's why it includes two other commits of you @pcarruscag . 

In the example FADO-file below I marked interesting spots with `TK::` and all caps

Maybe also intersting for @cvencro

### CODE REVIEW

**Comment by TobiKattmann** [comment]
This file can be tested with the data in https://github.com/su2code/TestCases/pull/63 and https://github.com/su2code/SU2/pull/1288/
```python
# Finite Differences of unsteady CHT

from FADO import *
import subprocess

# Design variables ----------------------------------------------------- #

# DV_VALUE of 18 DV's. And I need an additional all zero/baseline run.
nDV = 18
ffd = InputVariable(0.0,PreStringHandler("DV_VALUE="),nDV)

# Parameters ----------------------------------------------------------- #

# switch input mesh to perform deformation
# TK:: NOT NECESSARY ANYMORE
#mesh_in = Parameter(["MESH_FILENAME= mesh_out.su2"],\
#       LabelReplacer("MESH_FILENAME= MeshCHT.su2"))

# Evaluations ---------------------------------------------------------- #

def_command = "SU2_DEF chtMaster.cfg" # note that correct SU2 version needs to be in Path
cfd_command = "mpirun -n 12 SU2_CFD chtMaster.cfg"

max_tries = 1

# mesh deformation
deform = ExternalRun("DEFORM",def_command,True) # True means sym links are used for 
deform.setMaxTries(max_tries)
deform.addConfig("chtMaster.cfg")
deform.addConfig("testfile.txt")
deform.addData("fluid.cfg") # zonal cfg's can be symlinked
deform.addData("solid.cfg")
deform.addData("MeshCHT.su2")
deform.addExpected("mesh_out.su2")

# direct run
direct = ExternalRun("DIRECT",cfd_command,True) # TK:: WORKS WITH SYMLINKS AND HARDCOPYS
direct.setMaxTries(max_tries)
direct.addConfig("chtMaster.cfg")
direct.addData("fluid.cfg") # zonal cfg's can be symlinked
direct.addData("solid.cfg")
direct.addData("DEFORM/mesh_out.su2", destination="MeshCHT.su2") # TK:: DESTINATION ONLY WITH FILENAME WITHOUT FURTHER PATH
direct.addData("solution_0_00000.dat")
direct.addData("solution_0_00001.dat")
direct.addData("solution_1_00000.dat")
direct.addData("solution_1_00001.dat")
direct.addExpected("solution_0_00055.dat")
direct.addExpected("solution_1_00055.dat")
# TK:: NOT NECESSARY ANYMORE
#direct.addParameter(mesh_in)

# Functions ------------------------------------------------------------ #

tavgT = Function("tavgT", "DIRECT/chtMaster.csv",LabeledTableReader("\"tavg[AvgTemp[1]]\""))
tavgT.addInputVariable(ffd,"",None)
tavgT.addValueEvalStep(deform)
tavgT.addValueEvalStep(direct)

# Driver --------------------------------------------------------------- #

driver = ExteriorPenaltyDriver(0.005)
driver.addObjective("min", tavgT)

driver.setWorkingDirectory("DOE")
driver.preprocessVariables()
driver.setStorageMode(True,"DESIGN_")

his = open("doe.his","w",1)
driver.setHistorian(his)

# DOE
x = driver.getInitial()
driver.fun(x) # baseline evaluation

for iLoop in range(0, nDV, 1):
    x = driver.getInitial()
    x[iLoop] = 1e-4 # DV_VALUE
    driver.fun(x)
#end

his.close()


```

File: evaluation.py [hunk 1]
```diff
@@ -54,17 +55,22 @@ def _addAbsoluteFile(self,file,flist):
...
-                   is immediately converted to an absolute path, the file must exist),
-                   or "auto" (tries "absolute" first, reverts to "relative"). 
+        file       : Path to the file.
+        location   : Type of path, "relative" (to the parent of "dir"), "absolute" (the path
+                     is immediately converted to an absolute path, the file must exist),
+                     or "auto" (tries "absolute" first, reverts to "relative").
+        destination: Filename to be set at the destination. Discard any additional file path.
+                     Default destination is the regular filename
         """
+        if destination == None: destination = os.path.basename(file)
```
Comment (TobiKattmann) [inline evaluation.py]:
```suggestion
        if destination is None: destination = os.path.basename(file)
```

File: evaluation.py [hunk 2]
```diff
@@ -54,17 +55,22 @@ def _addAbsoluteFile(self,file,flist):
...
+        file       : Path to the file.
+        location   : Type of path, "relative" (to the parent of "dir"), "absolute" (the path
+                     is immediately converted to an absolute path, the file must exist),
+                     or "auto" (tries "absolute" first, reverts to "relative").
+        destination: Filename to be set at the destination. Discard any additional file path.
+                     Default destination is the regular filename
         """
+        if destination is None: destination = os.path.basename(file)
+        self._dataFilesDestination.append(destination)
+
```
Comment (pcarruscag) [inline evaluation.py]:
```suggestion
        if destination is None: destination = file
        self._dataFilesDestination.append(os.path.basename(destination))

```

File: evaluation.py [hunk 3]
```diff
@@ -119,8 +125,8 @@ def initialize(self):
 
         try:
             os.mkdir(self._workDir)
-            for file in self._dataFiles:
-                target = os.path.join(self._workDir,os.path.basename(file))
+            for file, destination in zip(self._dataFiles, self._dataFilesDestination):
+                target = os.path.join(self._workDir,os.path.basename(destination))
```
Comment (pcarruscag) [inline evaluation.py]:
```suggestion
                target = os.path.join(self._workDir,destination)
```

File: evaluation.py [hunk 4]
```diff
@@ -54,17 +55,22 @@ def _addAbsoluteFile(self,file,flist):
...
-        file     : Path to the file.
-        location : Type of path, "relative" (to the parent of "dir"), "absolute" (the path
-                   is immediately converted to an absolute path, the file must exist),
-                   or "auto" (tries "absolute" first, reverts to "relative"). 
+        file       : Path to the file.
+        location   : Type of path, "relative" (to the parent of "dir"), "absolute" (the path
+                     is immediately converted to an absolute path, the file must exist),
+                     or "auto" (tries "absolute" first, reverts to "relative").
+        destination: Filename to be set at the destination. Discard any additional file path.
+                     Default destination is the regular filename
```
Comment (pcarruscag) [inline evaluation.py]:
```suggestion
        file        : Path to the file.
        location    : Type of path, "relative" (to the parent of "dir"), "absolute" (the path
                     is immediately converted to an absolute path, the file must exist),
                     or "auto" (tries "absolute" first, falls back to "relative").
        destination : Filename to be set at the destination. Discards any additional file path.
                     The default destination is the regular filename (i.e. "file").
```

**Review by pcarruscag — COMMENTED** [review]
Good stuff Tobi :1st_place_medal: 
Please add yourself to AUTHORS.md (once I update develop)
And modify example3 (which has mesh deformation) to use this new feature.

File: evaluation.py (hunk 2 above)
Comment (TobiKattmann) [reply evaluation.py]: 👍 makes sense to have that logic in one place

**Comment by TobiKattmann** [comment]
>  Please add yourself to AUTHORS.md (once I update develop)

Ok as far as i see you updated your develop and I merged develop into my branch ... but the diff of this PR shows your changes (added Authors.md) although this PR is for your develop... 😕 I hope it did it correctly

And thanks for FADO altogether. I used to shabby bash scripts with rather basic and inflexible features so far and I can imagine making a lot of use of this! And the code is well written and understandable and especially the documentation with clear docstrings helped quite a bit wile looking around. Initially I was scratching my head a bit though with just the examples :) . Well done @pcarruscag (I should have used this earlier)

**Review by pcarruscag — APPROVED** [review]
I am glad someone else finds this useful :)
Congrats again on being the #1 contributor.

**Pull request merged**